### PR TITLE
Mark deprecated functionality with C++ attributes and resolve internal usage of deprecated functions

### LINKED ISSUE #4426: Remove deprecated code
All deprecated code should be removed.

List of deprecated items:

Data:
- SQLParser::parseSQLString
- RecordSet::totalRowCount
- SQLChannel::PROP_ASYNC

Foundation:
- AbstractObserver::accepts(Notification* pNf, const char* pName)
- DirectorySeparator::operator++ (postfix operator)
- Dynamic::Var::empty
- DynamicAny
- Event::Event(bool autoReset)
- FIFOEvent
- FIFOStrategy
- HashFunction
- HashStatistic
- HashTable
- NObserver::accepts(Notification* pNf, const char* pName)
- NumericString::uIntToStr
- Observer::accepts(Notification* pNf, const char* pName)
- SimpleHashTable

Net:
- HTTPResponse::HTTP_USEPROXY
- HTTPResponse::HTTP_REQUESTENTITYTOOLARGE
- HTTPResponse::HTTP_REQUESTURITOOLONG
- HTTPResponse::HTTP_UNSUPPORTEDMEDIATYPE
- NetworkInterface::NetworkInterfaceList
- Socket::select

Applications can define `POCO_SILENCE_DEPRECATED` to disable Poco deprecation warnings if needed.

### PR DESCRIPTION
**Pull request by matejk**
Closes #4426.

### CODE REVIEW

File: Foundation/include/Poco/ClassLibrary.h [hunk 1]
```diff
@@ -67,10 +67,10 @@ extern "C"	\
 #define POCO_BEGIN_MANIFEST_IMPL(fnName, base) \
 	bool fnName(Poco::ManifestBase* pManifest_)										\
 	{																				\
-		typedef base _Base;															\
-		typedef Poco::Manifest<_Base> _Manifest;									\
-		std::string requiredType(typeid(_Manifest).name());							\
-		std::string actualType(pManifest_->className());							\
+		using _Base = base;															\
```
Comment (aleks-f) [inline Foundation/include/Poco/ClassLibrary.h:70]: we should change this naming, underscore and capital first letter are [reserved for the implementation](https://devblogs.microsoft.com/oldnewthing/20230109-00/?p=107685)

**Review by aleks-f — CHANGES_REQUESTED** [review]
I am not sure if `@deprecated` is needed/used by `PocoDoc`, @obiltschnig would know

File: Foundation/include/Poco/LocalDateTime.h [hunk 2]
```diff
@@ -85,13 +85,13 @@ class Foundation_API LocalDateTime
 		/// Creates a LocalDateTime from the UTC time given in dateTime,
 		/// using the time zone differential of the current time zone.
 
-	//@ deprecated
```
Comment (matejk) [inline Foundation/include/Poco/LocalDateTime.h:88]: @obiltschnig, @aleks-f: Some ctors/functions in this class seem not to have an alternative to be used. They are for example used in Dynamic/Var.h.

Is complete class deprecated?

File: Crypto/include/Poco/Crypto/EVPPKey.h [hunk 3]
```diff
@@ -35,10 +35,8 @@
 namespace Poco {
 namespace Crypto {
 
-//@deprecated
-class ECKey;
-//@deprecated
-class RSAKey;
+class [[deprecated]] ECKey;
+class [[deprecated]] RSAKey;
```
Comment (matejk) [inline Crypto/include/Poco/Crypto/EVPPKey.h:39]: @obiltschnig , what shall be used instead of these?

Comment (aleks-f) [reply Crypto/include/Poco/Crypto/EVPPKey.h:39]: @matejk anything not going through EVP api is deprecated, see [this](https://github.com/pocoproject/poco/issues/3516#issuecomment-1100750944)

some serious breakage (anything non-EVP in the interfaces must go) to rectify that; I'm not sure we want to embark on that for 1.14

**Comment by matejk** [comment]
Functionality that was marked as deprecated with comments is now marked as such with C++ attribute with a few exceptions:

- Deprecating OpenSSL non-EVP functionality is moved to ticket #4602.
- Deprecating old MongoDB network protocol is moved to ticket #4603.

Usage of deprecated functions and classes in Poco code is replaced with non-deprecated variants.

Unit tests for deprecated functionality are disabled by default and can be enabled with `POCO_TEST_DEPRECATED`.

Some C++ code modernisation was done in the code while resolving deprecations.

@aleks-f, @obiltschnig

File: Crypto/include/Poco/Crypto/EVPPKey.h (hunk 3 above)
Comment (matejk) [reply Crypto/include/Poco/Crypto/EVPPKey.h]: Opened new ticket

File: Data/src/SQLChannel.cpp [hunk 4]
```diff
@@ -235,6 +232,7 @@ void SQLChannel::run()
 	{
 		try
 		{
+			sleepTime = 100;
```
Comment (aleks-f) [inline Data/src/SQLChannel.cpp:235]: warning: likely to cause merge conflict with https://github.com/pocoproject/poco/pull/4530

File: MongoDB/include/Poco/MongoDB/MessageHeader.h [hunk 5]
```diff
@@ -38,14 +38,24 @@ class MongoDB_API MessageHeader
 
 	enum OpCode
 	{
+#if false
```
Comment (aleks-f) [inline MongoDB/include/Poco/MongoDB/MessageHeader.h:41]: I guess this #if should be removed. 
Generally speaking, I wonder how much warnings this whole PR generates by default?

**Review by aleks-f — APPROVED** [review]
Overall looks good, than you. I'm just thinking it would be good to have a single place to disarm warnings somehow.

File: MongoDB/include/Poco/MongoDB/MessageHeader.h (hunk 5 above)
Comment (matejk) [reply MongoDB/include/Poco/MongoDB/MessageHeader.h:41]: Issue to resolve MongoDB deprecate functionality is #4603. That is a quite big task on its own.

The PR does not generate more warnings as code from devel, probably much less.

**Comment by matejk** [comment]
> Overall looks good, than you. I'm just thinking it would be good to have a single place to disarm warnings somehow.

Do you mean deprecation warnings that will be seen by the users of Poco?

**Comment by aleks-f** [comment]
> > Overall looks good, than you. I'm just thinking it would be good to have a single place to disarm warnings somehow.
> 
> Do you mean deprecation warnings that will be seen by the users of Poco? Do we want to disarm them?

I foresee some annoyance if we swamp the build with deprecation warnings. In any case, I think there should be an easy way to disable them.

**Comment by matejk** [comment]
> > > Overall looks good, than you. I'm just thinking it would be good to have a single place to disarm warnings somehow.
> > 
> > 
> > Do you mean deprecation warnings that will be seen by the users of Poco? Do we want to disarm them?
> 
> I foresee some annoyance if we swamp the build with deprecation warnings. In any case, I think there should be an easy way to disable them.

Introduced a macro `POCO_DEPRECATED` instead of using `[[deprecated]]` directly. An application can define `POCO_SILENCE_DEPRECATED` to have `POCO_DEPRECATED` defined to an empty define.

IMO the branch is ready to be merged to main.

@obiltschnig, @aleks-f, any objections?

**Pull request merged**